lib/fsm/follow.go

```diff
@@ -22,74 +22,107 @@ import (
 
 	"github.com/gravitational/gravity/lib/storage"
 
-	"github.com/gravitational/trace"
+	"github.com/cenkalti/backoff"
 	"github.com/sirupsen/logrus"
 )
 
 // GetPlanFunc is a function that returns an operation plan.
 type GetPlanFunc func() (*storage.OperationPlan, error)
 
 // PlanEvent represents an operation plan event.
-type PlanEvent interface{}
+type PlanEvent interface {
+	// IsPlanEvent returns true for events representing plan changes.
+	IsPlanEvent() bool
+}
 
-// PlanChanged is sent when plan phases change states.
-type PlanChanged struct {
+// PlanChangeEvent is sent when plan phases change states.
+type PlanChangeEvent struct {
 	// Change is a plan phase change.
 	Change storage.PlanChange
 	// Plan is the resolved operation plan.
 	Plan storage.OperationPlan
 }
 
-// PlanFinished is sent when the plan is fully completed or rolled back.
-type PlanFinished struct {
+// IsPlanEvent is for satisfying PlanEvent interface.
+func (e *PlanChangeEvent) IsPlanEvent() bool { return true }
+
+// PlanFinishEvent is sent when the plan is fully completed or rolled back.
+type PlanFinishEvent struct {
 	// Plan is the resolved operation plan.
 	Plan storage.OperationPlan
 }
 
+// IsPlanEvent is for satisfying PlanEvent interface.
+func (e *PlanFinishEvent) IsPlanEvent() bool { return true }
+
 // FollowOperationPlan returns a channel that receives phase updates for the
 // specified plan.
-func FollowOperationPlan(ctx context.Context, getPlan GetPlanFunc) (<-chan PlanEvent, error) {
-	previousPlan, err := getPlan()
+func FollowOperationPlan(ctx context.Context, getPlan GetPlanFunc) <-chan PlanEvent {
+	ch := make(chan PlanEvent, 100)
+	// Send an initial batch of events from the initial state of the plan.
+	plan, err := getPlan()
 	if err != nil {
-		return nil, trace.Wrap(err)
+		logrus.WithError(err).Error("Failed to load plan.")
 	}
-	ch := make(chan PlanEvent, 100)
-	go func() {
-		// Send initial events to the channel, excluding unstarted phases.
-		for _, change := range GetPlanProgress(*previousPlan) {
-			ch <- &PlanChanged{Plan: *previousPlan, Change: change}
+	if plan != nil {
+		for _, change := range GetPlanProgress(*plan) {
```

Comment:
this could potentially be encapsulated in a function and reused.
Also - why not pass the changes as a slice into the channel instead of one after another - it would make sense if changes were available separately but they are available all at once.